common/src/main/java/io/netty/util/concurrent/SingleThreadEventExecutor.java

```diff
@@ -979,81 +1093,115 @@ private boolean ensureThreadStarted(int oldState) {
-                        for (;;) {
-                            int oldState = state;
-                            if (oldState >= ST_SHUTDOWN || STATE_UPDATER.compareAndSet(
-                                    SingleThreadEventExecutor.this, oldState, ST_SHUTDOWN)) {
-                                break;
+                            // Now we want to make sure no more tasks can be added from this point. This is
+                            // achieved by switching the state. Any new tasks beyond this point will be rejected.
+                            for (;;) {
+                                int currentState = state;
+                                if (currentState >= ST_SHUTDOWN || STATE_UPDATER.compareAndSet(
+                                        SingleThreadEventExecutor.this, currentState, ST_SHUTDOWN)) {
+                                    break;
+                                }
                             }
-                        }
 
-                        // We have the final set of tasks in the queue now, no more can be added, run all remaining.
-                        // No need to loop here, this is the final pass.
-                        confirmShutdown();
+                            // We have the final set of tasks in the queue now, no more can be added, run all remaining.
+                            // No need to loop here, this is the final pass.
+                            confirmShutdown();
+                        }
                     } finally {
                         try {
-                            cleanup();
-                        } finally {
-                            // Lets remove all FastThreadLocals for the Thread as we are about to terminate and notify
-                            // the future. The user may block on the future and once it unblocks the JVM may terminate
-                            // and start unloading classes.
-                            // See https://github.com/netty/netty/issues/6596.
-                            FastThreadLocal.removeAll();
-
-                            STATE_UPDATER.set(SingleThreadEventExecutor.this, ST_TERMINATED);
-                            threadLock.countDown();
-                            int numUserTasks = drainTasks();
-                            if (numUserTasks > 0 && logger.isWarnEnabled()) {
-                                logger.warn("An event executor terminated with " +
-                                        "non-empty task queue (" + numUserTasks + ')');
+                            if (oldState == ST_SUSPENDED) {
+                                for (;;) {
+                                    if (!runAllTasks()) {
+                                        break;
+                                    }
+                                }
+                                // Lets remove all FastThreadLocals for the Thread as we are about to terminate it.
+                                FastThreadLocal.removeAll();
```

Comment:
Speaking of `FastThreadLocal`, does it deserve some additional documentation to point out the behavior with respect to loops being suspended?

Reply:
Can you clarify what you are after ? For me the comment above seems to be good enough

Reply:
I mean't adding something to the `FastThreadLocal` classes docs to cross reference the suspension behavior so people aren't surprised that even when using the same event loop their `FastThreadLocal` comes up empty.

Reply:
Not sure if this just confuse people more...  I think I would just pun on it for now